### PR TITLE
channel scaling spinboxes update

### PR DESCRIPTION
**Pull request by nmarkowitz**
#### Reference issue

https://github.com/mne-tools/mne-qt-browser/pull/230 , https://github.com/mne-tools/mne-python/issues/10888 , https://github.com/mne-tools/mne-qt-browser/pull/212

#### What does this implement/fix?

Spinboxes in channel scalings within Settings can now be updated

Screenshot 2024-07-02 at 2 05 50 PM

#### Additional information

This is the second of 4 or 5 PRs to allow display and scaling of channel types based on sensor units and monitor resolution and size

### CODE REVIEW

**Comment by larsoner** [comment]
@nmarkowitz just let me know when I should take a look / review / try it!

File: mne_qt_browser/_pg_figure.py [hunk 1]
```diff
@@ -1850,6 +1857,26 @@ def _value_changed(self, new_value, value_name):
...
+            )
+            self.mne.scalebar_texts[ch_type].update_value()
+            self.weakmain()._redraw()
+        else:
+            for ch_type, spinbox in self.ch_scaling_spinboxes.items():
+                spinbox.setValue(
+                    self.mne.scalings[ch_type]
+                    * self.mne.unit_scalings[ch_type]
+                    * scaler
+                    / self.mne.scale_factor
```
Comment (larsoner) [inline mne_qt_browser/_pg_figure.py:1877]: This...

File: mne_qt_browser/_pg_figure.py [hunk 2]
```diff
@@ -1809,19 +1809,26 @@ def __init__(self, main, title="Settings", **kwargs):
...
-                ch_spinbox.setValue(self.mne.scalings[ch] * self.mne.unit_scalings[ch])
-                ch_spinbox.setDisabled(True)
-                # ch_spinbox.setReadOnly(True)
-                ch_spinbox.setMinimumWidth(150)
+                ch_spinbox.setDecimals(1)
+                ch_spinbox.setValue(
+                    self.mne.scalings[ch]
+                    * self.mne.unit_scalings[ch]
+                    * scaler
+                    / self.mne.scale_factor
```
Comment (larsoner) [inline mne_qt_browser/_pg_figure.py:1827]: ... and this are identical (as are the `scaler` computations). Better to refactor into a helper I think as I commented on one of your commits

File: mne_qt_browser/_pg_figure.py [hunk 3]
```diff
@@ -1850,6 +1857,26 @@ def _value_changed(self, new_value, value_name):
...
         self.weakmain()._toggle_antialiasing()
 
+    def _update_spinbox_values(self, *args, **kwargs):
+        """Update spinbox values."""
+        scaler = 1 if self.mne.butterfly else 2
+        if len(args) > 0:
+            new_value = args[0]
+            ch_type = kwargs["ch_type"]
+            self.mne.scalings[ch_type] = new_value / (
+                self.mne.unit_scalings[ch_type] * scaler / self.mne.scale_factor
```
Comment (larsoner) [inline mne_qt_browser/_pg_figure.py:1867]: ... and this could use the same helper but with some divisions probably?

Comment (nmarkowitz) [reply mne_qt_browser/_pg_figure.py]: Added helper in 5eea6c6038e49fc44bf4dc85c3779f22c91d83c3

**Comment by larsoner** [comment]
@nmarkowitz when I change the channel scalings, the scale bar updates (good) but the channel scaling does not change -- the traces should get bigger and smaller as I click up and down:


https://github.com/mne-tools/mne-qt-browser/assets/2365790/b9e307a6-4eef-4fcd-97a8-68d89a4e2350



Compare this to when I use the existing +/- functionality (which appears to update properly):


https://github.com/mne-tools/mne-qt-browser/assets/2365790/d7c29013-e711-4873-913b-e7ef31b4c3f4



Do you see the same thing locally? If so, make sure you test interactively like this in addition to writing unit tests -- arguably it's a more important test to do anyway!

**Comment by larsoner** [comment]
Not for me -- here I'll change from 800 to 8000 which is a factor of 10, so the traces should get a factor of 10 smaller, but they don't change:

https://github.com/mne-tools/mne-qt-browser/assets/2365790/21cba3d3-edb7-4ebd-8f87-91d9463fa914

The traces get smaller/larger for you when you click the up/down arrows in the double spin box?

**Comment by nmarkowitz** [comment]
@larsoner it works for me


https://github.com/mne-tools/mne-qt-browser/assets/34498671/fcf96bfe-57e0-48b9-bca2-3ed04cb59459

Could you be using a different qt package such as PySide? I'm using PyQt5

**Comment by larsoner** [comment]
Yeah I'm using PyQt6. I just tried with PySide6 and observed the same thing. So maybe it's a Qt5 vs Qt6 thing? It's weird, not sure how to explain it.

File: mne_qt_browser/_pg_figure.py [hunk 4]
```diff
@@ -1809,19 +1815,22 @@ def __init__(self, main, title="Settings", **kwargs):
...
-        # Get all unique channel types
+        # Get all unique channel types and allow scaling
         ordered_types = self.mne.ch_types[self.mne.ch_order]
         unique_type_idxs = np.unique(ordered_types, return_index=True)[1]
         ch_types_ordered = [ordered_types[idx] for idx in sorted(unique_type_idxs)]
         for ch in ch_types_ordered:
             if ch in self.mne.unit_scalings.keys():
                 ch_spinbox = QDoubleSpinBox()
+                ch_spinbox.setMinimumWidth(100)
                 ch_spinbox.setRange(-float("inf"), float("inf"))
```
Comment (larsoner) [inline mne_qt_browser/_pg_figure.py:1826]: I don't think you should be able to set negative values. Zero is probably bad, too, but this is at least a step in the right direction:
```suggestion
                ch_spinbox.setRange(0, float("inf"))
```

File: mne_qt_browser/_pg_figure.py [hunk 5]
```diff
@@ -1809,19 +1815,22 @@ def __init__(self, main, title="Settings", **kwargs):
...
-                ch_spinbox.setDecimals(3)
-                ch_spinbox.setValue(self.mne.scalings[ch] * self.mne.unit_scalings[ch])
-                ch_spinbox.setDisabled(True)
-                # ch_spinbox.setReadOnly(True)
-                ch_spinbox.setMinimumWidth(150)
+                ch_spinbox.setDecimals(1)
```
Comment (larsoner) [inline mne_qt_browser/_pg_figure.py:1827]: Maybe 2 is better and setting to an adaptive step type could help?

https://doc.qt.io/qt-6/qdoublespinbox.html#stepType-prop

File: mne_qt_browser/_pg_figure.py [hunk 6]
```diff
@@ -1809,19 +1815,22 @@ def __init__(self, main, title="Settings", **kwargs):
...
+                inv_norm = _get_channel_scaling(self, ch)
+                ch_spinbox.setSingleStep(inv_norm * 0.25)
```
Comment (larsoner) [inline mne_qt_browser/_pg_figure.py:1829]: Have you tried adaptive, might work decently?

https://doc.qt.io/qt-6/qdoublespinbox.html#setStepType
https://doc.qt.io/qt-6/qdoublespinbox.html#stepType-prop

File: mne_qt_browser/_pg_figure.py [hunk 7]
```diff
@@ -3909,6 +3947,10 @@ def scale_all(self, checked=False, *, step):
 
         # Update Scalebars
         self._update_scalebar_values()
+        if self.mne.fig_settings is not None:
+            self.mne.fig_settings._update_spinbox_values()
+
+        # self._update_ch_spinbox_values()
```
Comment (larsoner) [inline mne_qt_browser/_pg_figure.py:3953]: Cruft, remove?

**Review by larsoner — COMMENTED** [review]
Just some minor comments.

For the PyQt stuff maybe try creating a conda env with `pyside6`?

**Comment by nmarkowitz** [comment]
@larsoner I just created a new environment that uses pyside6 and the spinboxes still work for me. I even tried running in a regular python console as opposed to an ipython terminal. Not sure why it isn't working for you

**Comment by larsoner** [comment]
Thanks @nmarkowitz !

**Pull request merged**